src/vidpf.rs

```diff
@@ -0,0 +1,743 @@
+        }
+
+        Ok(ValueShare {
+            share,
+            proof: state.proof,
+        })
+    }
+
+    /// [`Vidpf::eval_next`] evaluates the `input` at the level indicated by the provided state
+    /// and returns a new state and a share of the input's weight at the next level.
+    fn eval_next(
+        &self,
+        id: ServerId,
+        public: &PublicShare<W>,
+        input: &VidpfInput,
+        level: usize,
+        state: &EvalState,
+        nonce: &[u8; NONCE_SIZE],
+    ) -> Result<(EvalState, W), VidpfError> {
+        let CorrectionWord {
+            seed: ref s_cw,
+            left_control_bit: ref t_cw_l,
+            right_control_bit: ref t_cw_r,
+            weight: ref w_cw,
+        } = public.cw.get(level).ok_or(VidpfError::IndexLevel)?;
+
+        let seq_tilde = Self::prg(&state.seed, nonce);
+
+        let t_i = state.control_bit;
+        let sl = conditional_xor_seeds(&seq_tilde.left_seed, s_cw, t_i);
+        let sr = conditional_xor_seeds(&seq_tilde.right_seed, s_cw, t_i);
+        let tl = seq_tilde.left_control_bit ^ (t_i & *t_cw_l);
+        let tr = seq_tilde.right_control_bit ^ (t_i & *t_cw_r);
+
+        let x_i = Choice::from(u8::from(
+            input.get_checked(level).ok_or(VidpfError::IndexLevel)?,
+        ));
+        let s_tilde_i = conditional_select_seed(x_i, &[sl, sr]);
+
+        let next_control_bit = Choice::conditional_select(&tl, &tr, x_i);
+        let (next_seed, w_i) = self.convert(s_tilde_i, nonce);
+
+        let zero = <W as IdpfValue>::zero(&self.weight_parameter);
+        let mut y = w_i + <W as IdpfValue>::conditional_select(&zero, w_cw, next_control_bit);
+        y.conditional_negate(Choice::from(id));
+
+        let pi_i = &state.proof;
+        let cs_i = &public.cs.get(level).ok_or(VidpfError::IndexLevel)?;
+        let pi_tilde = Self::node_proof(input, level, &next_seed)?;
+        let h2_input = pi_tilde.conditional_xor(next_control_bit, cs_i).xor(pi_i);
+        let next_proof = Self::node_proof_adjustment(h2_input).xor(pi_i);
+
+        let next_state = EvalState {
+            seed: next_seed,
+            control_bit: next_control_bit,
+            proof: next_proof,
+        };
+
+        Ok((next_state, y))
+    }
+
+    fn prg(seed: &Seed, nonce: &[u8]) -> PrgOutput {
+        let mut rng = XofFixedKeyAes128::seed_stream(&XofSeed(*seed), DomainSepTag::PRG, nonce);
+
+        let mut left_seed = Seed::default();
+        let mut right_seed = Seed::default();
+        rng.fill_bytes(&mut left_seed);
+        rng.fill_bytes(&mut right_seed);
+        // Use the LSB of seeds as control bits, and clears the bit,
+        // i.e., seeds produced by `prg` always have their LSB = 0.
+        // This ensures `prg` costs two AES calls only.
+        let left_control_bit = Choice::from(left_seed[0] & 0x1);
+        let right_control_bit = Choice::from(right_seed[0] & 0x1);
```

Comment:
nit: Slightly more obvious which bit we're stealing (`0x01 ^ 0xfe == 0xff`)
```suggestion
        let left_control_bit = Choice::from(left_seed[0] & 0x01);
        let right_control_bit = Choice::from(right_seed[0] & 0x01);
```